File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.c
int lengthOfLongestSubstring(char* s) {
    int lastIndex[256];
    for (int i = 0; i < 256; i++) {
        lastIndex[i] = -1;  // -1 means character hasn't appeared yet
    }

    int maxLen = 0;
    int start = 0;  // left edge of current window

    for (int end = 0; s[end] != '\0'; end++) {
        unsigned char c = s[end];

        // If this character was seen inside the current window, shrink window
        if (lastIndex[c] >= start) {
            start = lastIndex[c] + 1;
        }

        lastIndex[c] = end;

        int currentLen = end - start + 1;
        if (currentLen > maxLen) {
            maxLen = currentLen;
        }
    }

    return maxLen;
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {

        boolean[] seen = new boolean[128]; 

        int left = 0;
        int maxLength = 0;

        for (int right = 0; right < s.length(); right++) {

            char current = s.charAt(right);
            while (seen[current]) {
                seen[s.charAt(left)] = false;
                left++;
            }

            seen[current] = true;
            maxLength = Math.max(maxLength, right - left + 1);
        }

        return maxLength;
    }
}